### PR TITLE
Arc-Length 1D Conv Decoder: translation-equivariant surface head

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

Map surface nodes to a canonical 1D arc-length grid (256 points), apply 1D depthwise-separable convolutions, then interpolate back. This gives **translation equivariance** along the airfoil surface — suction peaks can be detected regardless of where they appear on the arc-length. The current SRF MLP has no spatial inductive bias at all.

Different from spectral arc-length loss (#2288 closed, output-space regularization) and DCT freq loss (merged, frequency-domain loss). This changes the **architecture itself** to be spatially aware along the surface manifold.

### Step 1: Compute arc-length and interpolate to fixed grid

```python
def surface_to_grid(surf_feats, surf_xy, n_grid=256):
    """
    Interpolate irregular surface node features onto uniform arc-length grid.
    surf_feats: [B, N_surf, D]
    surf_xy: [B, N_surf, 2]
    Returns: grid_feats [B, D, n_grid], node_arc [B, N_surf]
    """
    # Compute arc-length for each node
    centroid = surf_xy.mean(dim=1, keepdim=True)
    angles = torch.atan2(surf_xy[:,:,1]-centroid[:,:,1], 
                          surf_xy[:,:,0]-centroid[:,:,0])
    sort_idx = angles.argsort(dim=1)
    
    # Sort XY by angle, compute cumulative arc-length
    sorted_xy = torch.gather(surf_xy, 1, sort_idx.unsqueeze(-1).expand_as(surf_xy))
    diffs = sorted_xy[:, 1:] - sorted_xy[:, :-1]
    seg_len = diffs.norm(dim=-1)
    arc = torch.zeros(surf_xy.shape[0], surf_xy.shape[1], device=surf_xy.device)
    arc[:, 1:] = torch.cumsum(seg_len, dim=1)
    arc = arc / (arc[:, -1:] + 1e-8)  # normalize to [0,1]
    
    # Sort features by angle too
    sorted_feats = torch.gather(surf_feats, 1, sort_idx.unsqueeze(-1).expand_as(surf_feats))
    
    # Interpolate to uniform grid using linear interpolation
    grid_coords = torch.linspace(0, 1, n_grid, device=arc.device)  # [n_grid]
    # For each grid point, find the two nearest arc-length nodes and interpolate
    # Use torch.searchsorted for efficient interpolation
    idx = torch.searchsorted(arc, grid_coords.unsqueeze(0).expand(arc.shape[0], -1))
    idx = idx.clamp(1, arc.shape[1]-1)
    
    alpha = (grid_coords.unsqueeze(0) - torch.gather(arc, 1, idx-1)) / \
            (torch.gather(arc, 1, idx) - torch.gather(arc, 1, idx-1) + 1e-8)
    alpha = alpha.clamp(0, 1).unsqueeze(-1)
    
    feat_lo = torch.gather(sorted_feats, 1, (idx-1).unsqueeze(-1).expand(-1,-1,sorted_feats.shape[-1]))
    feat_hi = torch.gather(sorted_feats, 1, idx.unsqueeze(-1).expand(-1,-1,sorted_feats.shape[-1]))
    grid_feats = feat_lo * (1-alpha) + feat_hi * alpha  # [B, n_grid, D]
    
    return grid_feats.permute(0,2,1), arc, sort_idx  # [B, D, n_grid]
```

### Step 2: 1D conv decoder

```python
class ArcConvDecoder(nn.Module):
    def __init__(self, d_in, d_out=3, n_grid=256, kernel_size=9):
        super().__init__()
        # 3-layer depthwise-separable 1D conv
        self.conv = nn.Sequential(
            # Depthwise
            nn.Conv1d(d_in, d_in, kernel_size, padding=kernel_size//2, groups=d_in),
            nn.SiLU(),
            nn.Conv1d(d_in, d_in, 1),  # pointwise
            nn.SiLU(),
            # Second layer
            nn.Conv1d(d_in, d_in, kernel_size, padding=kernel_size//2, groups=d_in),
            nn.SiLU(),
            nn.Conv1d(d_in, d_in, 1),
            nn.SiLU(),
            # Output projection
            nn.Conv1d(d_in, d_out, 1),
        )
    
    def forward(self, x):
        return self.conv(x)  # [B, d_out, n_grid]
```

### Step 3: Interpolate back to original node positions

After conv decoding on the grid, interpolate predictions back to original surface node positions using the arc-length coordinates (reverse of step 1).

### Step 4: Integration

Replace SRF MLP with: sort → interp to grid → 1D conv → interp back → unsort. Separate decoders for fore and aft foils.

### Flags
- `--srf_arc_conv` (bool)
- `--srf_arc_conv_points` (int, default=256)
- `--srf_arc_conv_kernel` (int, default=9)

### Training commands (2 seeds)
```bash
CUDA_VISIBLE_DEVICES=0 python train.py \
  --agent frieren --wandb_name "frieren/arc-conv-s42" --wandb_group arc-length-conv-decoder \
  --seed 42 --asinh_pressure --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --slice_num 96 --cosine_T_max 150 --pcgrad_3way \
  --pressure_first --pressure_deep --residual_prediction --surface_refine \
  --te_coord_frame --wake_deficit_feature --re_stratified_sampling --n_layers 3 \
  --cp_panel --cp_panel_tandem_only --cp_panel_scale 0.1 \
  --srf_arc_conv --srf_arc_conv_points 256 --srf_arc_conv_kernel 9

CUDA_VISIBLE_DEVICES=0 python train.py \
  --agent frieren --wandb_name "frieren/arc-conv-s73" --wandb_group arc-length-conv-decoder \
  --seed 73 --asinh_pressure --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --slice_num 96 --cosine_T_max 150 --pcgrad_3way \
  --pressure_first --pressure_deep --residual_prediction --surface_refine \
  --te_coord_frame --wake_deficit_feature --re_stratified_sampling --n_layers 3 \
  --cp_panel --cp_panel_tandem_only --cp_panel_scale 0.1 \
  --srf_arc_conv --srf_arc_conv_points 256 --srf_arc_conv_kernel 9
```

## Baseline
| Metric | Value | Target |
|--------|-------|--------|
| p_in | 11.709 | < 11.709 |
| p_oodc | 7.544 | < 7.544 |
| p_tan | 27.402 | < 27.402 |
| p_re | 6.481 | < 6.481 |

Reproduce: `cd cfd_tandemfoil && python train.py --asinh_pressure --field_decoder --adaln_output --use_lion --lr 2e-4 --slice_num 96 --cosine_T_max 150 --pcgrad_3way --pressure_first --pressure_deep --residual_prediction --surface_refine --te_coord_frame --wake_deficit_feature --re_stratified_sampling --n_layers 3 --cp_panel --cp_panel_tandem_only --cp_panel_scale 0.1`

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT frieren:

## Results

### Arc-Length 1D Conv Decoder (2-seed experiment)

| Metric | Seed 42 | Seed 73 | 2-seed avg | Baseline | Δ |
|--------|---------|---------|------------|----------|---|
| **p_in** | 12.1 | 12.3 | **12.20** | 11.709 | **+4.2%** ❌ |
| **p_oodc** | 7.7 | 7.5 | **7.60** | 7.544 | **+0.7%** ❌ |
| **p_tan** | 27.3 | 26.9 | **27.10** | 27.402 | **-1.1%** ✅ |
| **p_re** | 6.7 | 6.6 | **6.65** | 6.481 | **+2.6%** ❌ |

### Commands
```bash
# Seed 42
CUDA_VISIBLE_DEVICES=0 python train.py \
  --agent frieren --wandb_name "frieren/arc-conv-s42" --wandb_group arc-length-conv-decoder \
  --seed 42 --asinh_pressure --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --slice_num 96 --cosine_T_max 150 --pcgrad_3way \
  --pressure_first --pressure_deep --residual_prediction --surface_refine \
  --te_coord_frame --wake_deficit_feature --re_stratified_sampling --n_layers 3 \
  --cp_panel --cp_panel_tandem_only --cp_panel_scale 0.1 \
  --srf_arc_conv --srf_arc_conv_points 256 --srf_arc_conv_kernel 9

# Seed 73
CUDA_VISIBLE_DEVICES=1 python train.py \
  --agent frieren --wandb_name "frieren/arc-conv-s73" --wandb_group arc-length-conv-decoder \
  --seed 73 --asinh_pressure --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --slice_num 96 --cosine_T_max 150 --pcgrad_3way \
  --pressure_first --pressure_deep --residual_prediction --surface_refine \
  --te_coord_frame --wake_deficit_feature --re_stratified_sampling --n_layers 3 \
  --cp_panel --cp_panel_tandem_only --cp_panel_scale 0.1 \
  --srf_arc_conv --srf_arc_conv_points 256 --srf_arc_conv_kernel 9
```

### Run details
- **W&B runs:** h9e2woq3 (seed 42), dj590mup (seed 73)
- **Peak VRAM:** 40.5 GB (seed 42), 39.7 GB (seed 73)
- **Epochs:** 144 (seed 42), 143 (seed 73) — hit 180-min wall-clock limit
- **Arc-conv params:** 80,928 per head × 2 (fore + aft) = 161,856 total
- **Training speed:** ~75s/epoch (6× slower than baseline ~12s/epoch)

### What happened

**Net negative result.** The arc-length 1D conv decoder hurts in-distribution pressure prediction (p_in +4.2%) while providing only a modest improvement on tandem transfer (p_tan -1.1%). p_oodc and p_re also regress.

**Why it likely failed:**

1. **6× training slowdown.** The per-sample for loop in arc-conv (sorting, interpolation, unsorting per foil per sample) makes training ~6× slower. While both runs reached ~144 epochs (comparable to baseline's ~149), the cosine annealing schedule with T_max=150 was designed for ~149 epochs at ~12s/epoch. The 75s/epoch training means the learning rate schedule was the same in terms of epoch count, but the model saw each sample fewer effective gradient steps per wall-clock time.

2. **Information bottleneck at the 256-point grid.** The surface meshes have varying numbers of nodes (often >256), so the interpolation to a fixed 256-point grid smooths out fine-grained features that the MLP preserves. The suction peaks and pressure gradients at leading/trailing edges — exactly the features we need for p_in — may be attenuated.

3. **Arc-length parameterization is lossy for tandem.** The angle-based sorting from centroid works for single closed airfoils but is problematic for aft-foil surfaces where the geometry is more complex (wake interference region). This may explain why p_in (heavily influenced by aft-foil in tandem) regresses.

4. **Translation equivariance isn't the right inductive bias.** The hypothesis assumed suction peaks could be at arbitrary arc-length positions, but in practice suction peaks are highly correlated with geometry (leading edge position, curvature) which the MLP already captures through the hidden features. The conv's equivariance property doesn't add value because the model already has geometry-aware features.

### Suggested follow-ups
- If pursuing spatial inductive bias on surfaces, consider **attention-based** approaches (e.g., surface node self-attention) instead of fixed-grid convolutions — these wouldn't require the lossy interpolation step.
- A **smaller grid** (128 points) or **adaptive grid** that concentrates points at leading/trailing edges might address the information bottleneck.
- The 6× slowdown is a fundamental limitation of per-sample processing. Any future surface-aware decoder should operate in batched mode to be practical.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2343 because ADVISOR: Closing — Arc-length 1D conv decoder is a clear negative. p_in +4.2%, p_oodc +0.7%, p_re +2.6%, with only marginal p_tan -1.1%. The 6× training slowdown from per-sample for loops is a fundamental limitation. The 256-point grid interpolation smooths out fine-grained pressure features (LE suction peaks, TE gradients) that the MLP naturally preserves. Good diagnosis — translation equivariance is not the right inductive bias here because suction peaks are highly correlated with known geometry features, not arbitrary arc-length positions. Your suggestion about attention-based surface approaches is noted for future consideration..
